File: src/core/validation.ts
/**
 * Independent validators. Every success legality is recomputed from
 * explicit inputs rather than trusted from engine internals. That
 * includes the four hard-eligibility predicates, which are written out
 * again here on purpose: a checker that asked the engine whether a
 * candidate was legal could only agree with the engine.
 */

import type { ContextBundle } from "./bundle.ts";
import type {
  CompilationResult,
  ContextCandidate,
  ContextRequest,
  RequirementClass,
} from "./domain.ts";
import type { CompilerPolicy } from "./policy.ts";

/**
 * The hard-eligibility rules, restated independently of the engine.
 * Returns the name of the first failed rule, or null when legal.
 */
function illegality(c: ContextCandidate): string | null {
  if (c.scopeEligible !== true) return "scope";
  if (c.freshnessEligible !== true) return "freshness";
  if (c.authorityEligible !== true) return "authority";
  if (c.formRank < c.minRank) return "floor";
  return null;
}

export function validateBundle(
  bundle: ContextBundle,
  request: ContextRequest,
  candidates: readonly ContextCandidate[],
  _policy: CompilerPolicy,
): string[] {
  void _policy;
  const problems: string[] = [];
  const byId = new Map(candidates.map((c) => [c.candidateId, c]));
  const liveIds = bundle.items.map((item) => item.id);

  if (
    liveIds.length !== bundle.layoutTrace.length ||
    liveIds.some((id, index) => id !== bundle.layoutTrace[index])
  ) {
    problems.push("layout trace must equal item order exactly");
  }

  const total = bundle.items.reduce((sum, item) => sum + item.tokenCount, 0);
  if (total > request.usableTokenBudget) {
    problems.push(
      `rendered cost ${total} exceeds budget ${request.usableTokenBudget}`,
    );
  }

  for (const itemId of liveIds) {
    const candidate = byId.get(itemId);
    if (!candidate) {
      problems.push(`bundle item not a candidate: ${itemId}`);
      continue;
    }
    const failed = illegality(candidate);
    if (failed !== null) {
      problems.push(`admitted hard-ineligible candidate ${itemId}: ${failed}`);
    }
    if (candidate.formRank < candidate.minRank) {
      problems.push(`admitted candidate below floor: ${itemId}`);
    }
  }

  const requiredIds = new Set(request.requiredIds);
  const effective = (c: ContextCandidate): RequirementClass => {
    if (requiredIds.has(c.candidateId)) {
      return c.requirement === "MANDATORY" ? "MANDATORY" : "REQUIRED";
    }
    return c.requirement;
  };
  for (const requiredId of request.requiredIds) {
    if (!liveIds.includes(requiredId)) {
      const candidate = byId.get(requiredId);
      const band = candidate ? effective(candidate) : null;
      if (band === "MANDATORY" || band === "REQUIRED") {
        problems.push(`explicitly required candidate absent: ${requiredId}`);
      }
    }
  }

  for (const itemId of liveIds) {
    const candidate = byId.get(itemId);
    if (!candidate) continue;
    for (const dep of candidate.dependsOn) {
      if (!liveIds.includes(dep)) {
        problems.push(`unsatisfied dependency: ${itemId} needs ${dep}`);
      }
    }
  }

  const groups = new Map<string, string[]>();
  for (const candidate of candidates) {
    if (candidate.groupId && candidate.groupRequired) {
      const list = groups.get(candidate.groupId) ?? [];
      list.push(candidate.candidateId);
      groups.set(candidate.groupId, list);
    }
  }
  for (const [groupId, members] of groups) {
    const present = members.filter((mid) => liveIds.includes(mid));
    if (present.length > 0 && present.length !== members.length) {
      problems.push(`required group partially admitted: ${groupId}`);
    }
  }

  const seenContent = new Set<string>();
  for (const itemId of liveIds) {
    const candidate = byId.get(itemId);
    if (!candidate) continue;
    if (seenContent.has(candidate.contentIdentity)) {
      problems.push(
        `duplicate representation admitted: ${candidate.contentIdentity}`,
      );
    }
    seenContent.add(candidate.contentIdentity);
  }
  return problems;
}

export function validateResult(
  result: CompilationResult,
  request: ContextRequest,
  candidates: readonly ContextCandidate[],
): string[] {
  const problems: string[] = [];
  const traced = new Set(result.trace.entries.map((e) => e.candidateId));
  const considered = new Set(candidates.map((c) => c.candidateId));
  if (result.success) {
    if (
      traced.size !== considered.size ||
      [...considered].some((id) => !traced.has(id))
    ) {
      problems.push(
        `trace covers ${traced.size} of ${considered.size} considered candidates`,
      );
    }
  } else if ([...traced].some((id) => !considered.has(id))) {
    problems.push("trace references unconsidered candidates");
  }
  if (result.trace.requestId !== request.requestId) {
    problems.push("trace request mismatch");
  }
  if (result.success && result.failure !== null) {
    problems.push("success carries a failure");
  }
  if (!result.success && result.failure === null) {
    problems.push("failure carries no failure record");
  }
  if (!result.success && result.bundleId !== null) {
    problems.push("failure carries a bundle identity");
  }
  if (result.trace.legacySchema !== "v1") {
    problems.push(...traceSemantics(result, candidates));
  }
  return problems;
}

/** Transitive dependencies in discovery order, excluding the start. */
function transitiveDependencies(
  start: string,
  byId: ReadonlyMap<string, ContextCandidate>,
): string[] {
  const seen = new Set<string>([start]);
  const order: string[] = [];
  const visit = (id: string): void => {
    for (const dep of byId.get(id)?.dependsOn ?? []) {
      if (seen.has(dep)) continue;
      seen.add(dep);
      order.push(dep);
      visit(dep);
    }
  };
  visit(start);
  return order;
}

/**
 * v2 trace semantics: an admission must show the budget it spent and,
 * where the candidate has dependencies, the closure that came with it.
 * These are exactly the two fields a v1 trace left meaningless.
 */
function traceSemantics(
  result: CompilationResult,
  candidates: readonly ContextCandidate[],
): string[] {
  const problems: string[] = [];
  const byId = new Map(candidates.map((c) => [c.candidateId, c]));
  for (const entry of result.trace.entries) {
    const candidate = byId.get(entry.candidateId);
    if (!candidate) continue;
    const expectedClosure = transitiveDependencies(entry.candidateId, byId);
    if (
      entry.dependencyClosure.length !== expectedClosure.length ||
      entry.dependencyClosure.some((id, i) => id !== expectedClosure[i])
    ) {
      problems.push(`trace closure wrong for ${entry.candidateId}`);
    }
    if (entry.decision === "ADMITTED") {
      if (entry.budgetBefore - entry.budgetAfter < candidate.tokenCount) {
        problems.push(`trace budget did not advance for ${entry.candidateId}`);
      }
    } else if (entry.budgetAfter > entry.budgetBefore) {
      problems.push(`trace budget rose for ${entry.candidateId}`);
    }
    if (entry.pulledInBy.length > 0 && entry.decision !== "ADMITTED") {
      problems.push(`trace pulled-in-by on non-admission ${entry.candidateId}`);
    }
  }
  return problems;
}

File: src/context_compiler/validation.py
"""Explicit validation for successful bundles and compilation results.

Independent of the engine internals that "probably" produced a legal
result: every check below recomputes legality from the request,
candidates, policy, bundle, and trace. A successful bundle must pass
all of them; anything else is a validator problem list, never an
exception for normal infeasibility (use CompileFailure for that).
"""

from __future__ import annotations

from context_compiler.bundle import ContextBundle
from context_compiler.domain import (
    CompilationResult,
    ContextCandidate,
    ContextRequest,
    RequirementClass,
    TraceDecision,
)
from context_compiler.engine import eligibility
from context_compiler.policy import CompilerPolicy


def validate_bundle(
    bundle: ContextBundle,
    request: ContextRequest,
    candidates: list[ContextCandidate],
    policy: CompilerPolicy,
) -> list[str]:
    """Recompute success legality from explicit inputs. Empty = valid."""
    problems: list[str] = []
    by_id = {c.candidate_id: c for c in candidates}
    live_ids = [item.id for item in bundle.items]

    if list(bundle.layout_trace) != live_ids:
        problems.append("layout trace must equal item order exactly")

    total = bundle.rendered_token_total()
    if total > request.usable_token_budget:
        problems.append(
            f"rendered cost {total} exceeds budget {request.usable_token_budget}"
        )

    for item_id in live_ids:
        candidate = by_id.get(item_id)
        if candidate is None:
            problems.append(f"bundle item not a candidate: {item_id}")
            continue
        eligible, code, _ = eligibility(candidate)
        if not eligible:
            problems.append(f"admitted hard-ineligible candidate {item_id}: {code}")
        if candidate.form_rank < candidate.min_rank:
            problems.append(f"admitted candidate below floor: {item_id}")

    effective = {
        c.candidate_id: (
            RequirementClass.MANDATORY
            if c.candidate_id in set(request.required_ids)
            and c.requirement == RequirementClass.MANDATORY
            else (
                RequirementClass.REQUIRED
                if c.candidate_id in set(request.required_ids)
                else c.requirement
            )
        )
        for c in candidates
    }
    _ = policy
    for required_id in request.required_ids:
        if required_id not in live_ids:
            candidate = by_id.get(required_id)
            band = effective.get(required_id)
            if candidate is not None and band in (
                RequirementClass.MANDATORY,
                RequirementClass.REQUIRED,
            ):
                problems.append(f"explicitly required candidate absent: {required_id}")

    for item_id in live_ids:
        candidate = by_id.get(item_id)
        if candidate is None:
            continue
        for dep in candidate.depends_on:
            if dep not in live_ids:
                problems.append(f"unsatisfied dependency: {item_id} needs {dep}")

    groups: dict[str, list[str]] = {}
    for candidate in candidates:
        if candidate.group_id and candidate.group_required:
            groups.setdefault(candidate.group_id, []).append(candidate.candidate_id)
    for group_id, members in groups.items():
        present = [mid for mid in members if mid in live_ids]
        if present and len(present) != len(members):
            problems.append(f"required group partially admitted: {group_id}")

    seen_content: set[str] = set()
    for item_id in live_ids:
        candidate = by_id.get(item_id)
        if candidate is None:
            continue
        if candidate.content_identity in seen_content:
            problems.append(
                f"duplicate representation admitted: {candidate.content_identity}"
            )
        seen_content.add(candidate.content_identity)

    return problems


def validate_result(
    result: CompilationResult,
    request: ContextRequest,
    candidates: list[ContextCandidate],
) -> list[str]:
    """Validate a compilation result envelope. Empty = valid."""
    problems: list[str] = []
    traced = {entry.candidate_id for entry in result.trace.entries}
    considered = {c.candidate_id for c in candidates}
    if result.success:
        # Every considered candidate gets exactly one terminal entry.
        if traced != considered:
            problems.append(
                f"trace covers {len(traced)} of {len(considered)} considered candidates"
            )
    elif not traced <= considered:
        # Early failures trace only candidates considered so far, but
        # must never invent entries for unseen candidates.
        problems.append("trace references unconsidered candidates")
    for entry in result.trace.entries:
        if not isinstance(entry.decision, TraceDecision):
            problems.append(f"non-terminal trace decision: {entry.candidate_id}")
    if result.trace.request_id != request.request_id:
        problems.append("trace request mismatch")
    if result.success and result.failure is not None:
        problems.append("success carries a failure")
    if not result.success and result.failure is None:
        problems.append("failure carries no failure record")
    if not result.success and result.bundle_id is not None:
        problems.append("failure carries a bundle identity")
    return problems
